### DISCUSSION TITLE
Question: automations, locks with toolkit

### DISCUSSION THREAD

**Discussion by anarcher** [fluxcd/flux2#220] [Q&A] [answered]
Hello,

First, I want to thank you for your contribution to the GitOps Ecosystem.
I am new to gitops toolkit. does the toolkit support automations and locks like fluxcd v1? and can I diff the changes before releasing via CLI or API?

Thanks

**Comment by squaremo** [comment, marked as answer]
There's [an implementation](https://github.com/fluxcd/image-automation-controller) and a design discussion -- https://github.com/fluxcd/toolkit/discussions/107.

The purpose of the controller is to do image update automation, but it works in a different way to Flux v1, because it follows the conventions of using custom resources and running individual controllers and so on. The implementation does the minimum, so far -- the design discussion is about a more capable and convenient design.

**Reply by anarcher** [reply]
Thanks for your reply.  I'm not sure but some changes (not only image changes) in git can diff or be confirmed before applying if operators/developers want it. and locking too. ( `KustomizePolicy` like `ImagePolicy`?)

**Reply by squaremo** [reply]
Regarding images:

You can effectively lock an image policy by setting it to an exact version. This differs a little from Flux v1, where you lock a _workload_ at its current image versions, to temporarily stop automation updating it. But that should also be possible by making changes to either the automation object (in the design, this is what switches automation on for certain files in a git repo), or to the file for the workload.

Another use of "lock" is to prevent accidental changes. For example, if there's a "poison" image that you don't want to be rolled out, you might lock the workload using it, preventing both automation and user interface from doing so. I think this is a legitimate use case that's not quite covered by the design so far.

Regarding kustomizations:

With the kustomization controller, there's no separation of source (i.e., git repository) and policy (i.e., version). You can "lock" a GitRepository at a particular version, meaning it won't go past that -- but that locks every kustomization object using that git repository. It's an issue I intend to bring up in a discussion of its own.
